core/src/main/java/net/kyori/moonshine/MoonshineBuilder.java

```diff
@@ -0,0 +1,239 @@
+    private final Map<Type, NavigableSet<Weighted<? extends IPlaceholderResolver<? extends R, ?, ? extends F>>>>
+        weightedPlaceholderResolvers = new HashMap<>();
+
+    private Resolved(final TypeToken<T> proxiedType,
+        final NavigableSet<Weighted<? extends IReceiverLocatorResolver<? extends R>>> weightedReceiverLocatorResolvers,
+        final IMessageSource<R, I> messageSource, final IMessageRenderer<R, I, O, F> messageRenderer,
+        final IMessageSender<R, O> messageSender,
+        final IPlaceholderResolverStrategy<R, I, F> placeholderResolverStrategy) {
+      this.proxiedType = proxiedType;
+      this.weightedReceiverLocatorResolvers = weightedReceiverLocatorResolvers;
+      this.messageSource = messageSource;
+      this.messageRenderer = messageRenderer;
+      this.messageSender = messageSender;
+      this.placeholderResolverStrategy = placeholderResolverStrategy;
+    }
+
+    @Deterministic
+    public <Z> @This Resolved<T, R, I, O, F> weightedPlaceholderResolver(
+        final Class<? extends Z> resolvedType,
+        final Weighted<? extends IPlaceholderResolver<? extends R, ? super Z, ? extends F>> weightedPlaceholderResolver) {
+      this.weightedPlaceholderResolvers.computeIfAbsent(resolvedType, ignored -> new TreeSet<>())
+          .add(weightedPlaceholderResolver);
+      return this;
+    }
+
+    @Deterministic
+    public <Z> @This Resolved<T, R, I, O, F> weightedPlaceholderResolver(
+        final TypeToken<? extends Z> resolvedType,
+        final Weighted<? extends IPlaceholderResolver<? extends R, ? super Z, ? extends F>> weightedPlaceholderResolver) {
+      this.weightedPlaceholderResolvers.computeIfAbsent(resolvedType.getType(), ignored -> new TreeSet<>())
+          .add(weightedPlaceholderResolver);
+      return this;
+    }
+
+    @Deterministic
+    public <Z> @This Resolved<T, R, I, O, F> weightedPlaceholderResolver(
+        final Class<? extends Z> resolvedType,
+        final IPlaceholderResolver<? extends R, ? super Z, ? extends F> placeholderResolver,
+        final int weight) {
+      this.weightedPlaceholderResolvers.computeIfAbsent(resolvedType, ignored -> new TreeSet<>())
+          .add(new Weighted<>(placeholderResolver, weight));
+      return this;
+    }
+
+    @Deterministic
+    public <Z> @This Resolved<T, R, I, O, F> weightedPlaceholderResolver(
+        final TypeToken<? extends Z> resolvedType,
+        final IPlaceholderResolver<? extends R, ? super Z, ? extends F> placeholderResolver,
+        final int weight) {
+      this.weightedPlaceholderResolvers.computeIfAbsent(resolvedType.getType(), ignored -> new TreeSet<>())
+          .add(new Weighted<>(placeholderResolver, weight));
+      return this;
+    }
+
+    @SideEffectFree
+    public T create() throws UnscannableMethodException {
+      return this.create(Thread.currentThread().getContextClassLoader());
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
Missing explanation comment (I think a "this is expected to fail if it cannot cast successfully" comment should be sufficient)